wg-api/spec-documents/message-ports.md

```diff
@@ -0,0 +1,206 @@
+
+* `channel` String
+* `message` any
+* `options` Object (optional)
+  * `transfer` any[] - a list of transferable objects to be sent with the message.
+
+This behaves similarly to `window.postMessage`, but instead of sending a message to another window, it sends the message to the main process.
+
+Also, this behaves similarly to `ipcRenderer.send`, but only takes a single argument after the channel (the message to be sent), and allows for additional options to be specified. In particular, it allows specifying a list of transferable objects to be sent along with the message.
+
+When this message is received in the main process, any `MessagePort` objects transferred will be deserialized as `ipcMain.MessagePort`.
+
+#### Class: WebContents
+
+##### `WebContents.postMessage(channel, message[, options])`
+
+* `channel` String
+* `message` any
+* `options` Object (optional)
+  * `transfer` any[] - a list of transferable objects to be sent with the message.
+
+Similar to `WebContents.send`, but allows transfering transferable objects.
+
+> [name=Samuel Attard]
+> I'm trying to figure out if there's a way for this to be non-breaking integrated into `.send` to avoid the confusion around two different ways to send IPC.  Can't see an easy way out there though.
+
+#### Class: ipcMain.MessageChannel extends EventEmitter
+
+Analogous to Blink's `MessageChannel`, but outside of Blink. Allows creating MessageChannels in the main process.
+
+##### Instance Properties
+
+* `port1` ipcMain.MessagePort
+* `port2` ipcMain.MessagePort
+
+#### Class: ipcMain.MessagePort
+
+Analogous to Blink's `MessagePort`, but outside of Blink. If sent to a renderer process, deserializes as a Blink `MessagePort`.
+
+##### Instance Methods
+
+###### `MessagePort.postMessage(message[, transferList])`
+
+See [Web Messaging](https://html.spec.whatwg.org/multipage/web-messaging.html#dom-messageport-postmessage), [Node](https://nodejs.org/dist/latest-v12.x/docs/api/worker_threads.html#worker_threads_port_postmessage_value_transferlist).
+
+###### `MessagePort.start()`
+
+See [Web Messaging](https://html.spec.whatwg.org/multipage/web-messaging.html#dom-messageport-start), [Node](https://nodejs.org/dist/latest-v12.x/docs/api/worker_threads.html#worker_threads_port_start)
+
+###### `MessagePort.close()`
+
+See [Web Messaging](https://html.spec.whatwg.org/multipage/web-messaging.html#dom-messageport-close), [Node](https://nodejs.org/dist/latest-v12.x/docs/api/worker_threads.html#worker_threads_port_close)
+
+##### Instance Events
+
+###### Event: 'message'
+
+Emitted when the port receives a message.
+
+###### Event: 'close'
+
+Emitted when the channel is closed.
+
+::::warning
+:warning: This event is not part of the Web Messaging API.
+::::
+
+> [name=Samuel Attard]
+> Is there a way to tell if a message port has been closed _after_ the fact.  E.g. `port.isClosed` or something?  Otherwise I assume `postMessage` will throw if the port is closed and you try to send a message.
```

Comment:
nope. none of the above, i believe.

Reply:
I propose we patch in an `onclose` event. It's pretty straightforward code-wise: https://github.com/electron/electron/pull/22532 and has some indicators of support from browsers: https://github.com/w3ctag/design-reviews/issues/269#issuecomment-407584290

Reply:
MessagePorts don't begin receiving events (including the "closed" event) until `.start()` is called on them. There's no race condition here.